Allow switching on and off individual tracers via cli

rvl opened the pull request:
Relates to #1235 

# Overview

- Adds structured logging in more places
- Allow switching on or off individual tracers via cli options.
- Change the type of logging from Trace -> Tracer.

# Comments

```
$ cardano-wallet-jormungandr launch --help-tracing 
Additional tracing options:

  --log-level SEVERITY     Global minimum severity for a message to be logged.
                           Defaults to "DEBUG" unless otherwise configured.
  --trace-NAME=off         Disable logging on the given tracer.
  --trace-NAME=SEVERITY    Set the minimum logging severity for the given
                           tracer. Defaults to "INFO".

The possible log levels (lowest to highest) are:
  debug info notice warning error critical alert emergency

The possible tracers are:
  application  
  api-server  
  api-worker  
  wallet-databases  
  wallet-db  
  network  
  stake-pool-monitor  
  stake-pool-layer  
  stake-pool-db  
  daedalus-ipc  
  worker-registry  
```

rvl committed:
Adding CLI options for enabling individual tracers

- CLI options can set severity or disable tracers for each component.
- More structured logging.
- Change Trace -> Tracer
- Use cardano-node style bundle of Trace m.

KtorZ commented on lib/core/src/Cardano/Wallet/DaedalusIPC.hs:
```diff
@@ -109,36 +123,66 @@ instance ToJSON MsgOut where
+        logTrace tr MsgSendHello
         pure Started
 
+    msg' m = logTrace tr (MsgReceive m) >> msg m
```
**[suggestion]** 

```hs
runNodeChannel hello (withLog msg)
{- ... -}
withLog action m = logTrace tr (MsgReceive m) >> action m
```

**Why?** 

Avoid cryptic `msg'` here and convey more meaning in the name of the function at the call site.

rvl replied:
OK good idea.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -54,10 +56,7 @@ module Cardano.CLI
-    , Verbosity (..)
     , withLogging
-    , verbosityToArgs
-    , verbosityToMinSeverity
```
**[remark]**

These functions should probably be removed completely ?

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs:
```diff
@@ -447,3 +427,69 @@ exitCodeApiServer = \case
+    , stakePoolLayerTracer :: f StakePoolLayerLog
+    , dbLayerTracer :: f DBLog
+    , daedalusIPCTracer :: f DaedalusIPCLog
+    }
```
**[remark]**

I am not sure to get the choice of names here :thinking: ? 
`dbLayer` refers to both the stake pools database and the wallet database (which we might want to distinguish I suppose). Also, the logs from the network seems missing, oversight or on purpose ?

rvl replied:
I revised the list of tracers.

KtorZ reviewed (commented):
looks good so far :+1: 
Nice to see these pieces put together.

KtorZ committed:
add descriptions to tracers

- Also removed the unused 'apiWorkerTracer'
- Also grouped 'walletDatabasesTracer' within 'applicationTracer'

KtorZ committed:
pad tracer descriptions for readability

KtorZ committed:
relocate WorkerRegistryLog and group worker logs in stake pool layer

KtorZ reviewed (approved):
LGTM. Played a bit with it, completed missing description for the tracers and grouped some tracers together to make the interface a bit clearer :+1: 

Nice work.

KtorZ committed:
adjust integration tests to match new logging

- There's no more --quiet nor --verbose options.
- The launcher now have specific error codes for various start-up errors

KtorZ committed:
remove now unused dependency from core.cabal
